src/test_fixture/world.rs

```diff
@@ -108,15 +136,54 @@ impl TestWorld {
     #[must_use]
     pub fn contexts<F: Field>(&self) -> [SemiHonestContext<'_, F>; 3] {
         let execution = self.executions.fetch_add(1, Ordering::Release);
-        let run = format!("run-{execution}");
-        zip(Role::all(), zip(&self.participants, &self.gateways))
+        zip(Role::all(), zip(&self.participants, &*self.gateways))
             .map(|(role, (participant, gateway))| {
-                SemiHonestContext::new(*role, participant, gateway).narrow(&run)
+                SemiHonestContext::new(*role, participant, gateway)
+                    .narrow(&Self::execution_step(execution))
             })
             .collect::<Vec<_>>()
             .try_into()
             .unwrap()
     }
+
+    #[must_use]
+    pub fn metrics_snapshot(&self) -> Metrics {
+        self.metrics_handle.snapshot()
+    }
+
+    #[must_use]
+    pub fn execution_step(execution: usize) -> impl Substep {
+        format!("run-{execution}")
+    }
+
+    pub fn gateway(&self, role: Role) -> &Gateway {
+        &self.gateways[role]
+    }
+
+    #[cfg(not(feature = "shuttle"))]
+    pub async fn join(mut self) {
+        // SAFETY: self is consumed by this method, so nobody can access gateways field after
+        // calling this method.
+        // joined flag is used inside the destructor to avoid double-free
```

Comment:
I'm not sure that this is safe unless accesses of `self.joined` are made thread-safe (atomic release-acquire or a mutex).  Thoughts?

Reply:
I tried to figure out how exactly it can be broken - both drop and join require ownership, so I am having hard time understanding how they can be called simultaneously. Having said that, atomic is fine and I'll just add it to be extra safe